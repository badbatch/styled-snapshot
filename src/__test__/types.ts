import { Func } from "@repodog/types";
import { ComponentType } from "enzyme";
import { ReactElement, ReactFragment, ReactPortal, Ref } from "react";
import { FlattenSimpleInterpolation } from "styled-components";
import { ContextConsumerElement, ContextProviderElement, ForwardRefElement, MemoElement } from "../types";

export interface ComponentProps {
  callback?: Func;
  Component?: ComponentType<{}>;
  consumer?: ContextConsumerElement;
  css?: FlattenSimpleInterpolation;
  decorator?: ForwardRefElement;
  element?: ReactElement;
  forwardRef?: ForwardRefElement;
  fragment?: ReactFragment;
  innerRef?: Ref<HTMLDivElement>;
  memo?: MemoElement;
  portal?: ReactPortal;
  provider?: ContextProviderElement;
  renderProp?: (...args: any[]) => ReactElement;
  styled?: ComponentType<{}>;
  styledFunctionComponent?: ReactElement;
  styledStyledComponent?: ReactElement;
  styles?: FlattenSimpleInterpolation;
  value?: string;
}
