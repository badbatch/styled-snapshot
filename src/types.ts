import { Func, ObjectMap } from "@repodog/types";
import { ShallowWrapper } from "enzyme";
import {
  ComponentClass,
  ComponentType,
  Consumer,
  ConsumerProps,
  ExoticComponent,
  ForwardRefExoticComponent,
  FunctionComponent,
  MemoExoticComponent,
  PropsWithChildren,
  PropsWithoutRef,
  Provider,
  ProviderProps,
  ReactElement,
  ReactNode,
  ReactPortal,
  RefAttributes,
  RefForwardingComponent,
} from "react";
import { AnyStyledComponent } from "styled-components";

export type ComponentTypeElement =
  | ReactElement<PropsWithChildren<{}>, ComponentClass>
  | ReactElement<PropsWithChildren<{}>, FunctionComponent>;

export type ContextConsumerElement = ReactElement<ConsumerProps<any>, Consumer<any>>; // tslint:disable-line no-any

// tslint:disable-next-line no-any
export type ContextProviderElement = ReactElement<ProviderProps<any>, Provider<any> & { _context: PrivateContext }>;

export type DomElement = ReactElement<PropsWithChildren<{}>, string>;

export type ForwardRefComponent = ForwardRefExoticComponent<
  PropsWithoutRef<Element> & RefAttributes<PropsWithChildren<{}>>
> & {
  readonly $$typeof: symbol;
  render: RefForwardingComponent<Element, PropsWithChildren<{}>>;
};

export type ForwardRefElement = ReactElement<PropsWithChildren<{}>, ForwardRefComponent>;

export type FragmentElement = ReactElement<PropsWithChildren<{}>, ExoticComponent<{ children?: ReactNode }>>;

export type MandatoryUnwrapElement =
  | ContextConsumerElement
  | ContextProviderElement
  | ForwardRefElement
  | FragmentElement
  | MemoElement
  | MemoExoticComponent<ComponentType>
  | ReactPortal;

export type MemoElement = ReactElement<PropsWithChildren<{}>, MemoExoticComponent<ComponentType>>;

export interface PrivateContext {
  Consumer: Consumer<any>; // tslint:disable-line no-any
  Provider: Provider<any>; // tslint:disable-line no-any
}

export type ValidElement = ComponentTypeElement | MandatoryUnwrapElement | DomElement;

export type ReactTreeVisitor = (node: ObjectMap) => void;

export interface StyledSnapshotConfig {
  elementsToUnwrap?: string[];
  reactTreeVisitor?: ReactTreeVisitor;
  unwrapCustomizer?: UnwrapCustomizer;
}

export type UnwrapCustomizer = (element: ValidElement) => ValidElement | ReactNode;

export type ComponentTree = ShallowWrapper<
  {
    children?: ReactNode;
  },
  Readonly<{}>,
  React.Component<{}, {}, any> // tslint:disable-line no-any
>;

export interface SerializedTree {
  $$typeof: symbol;
  children: SerializedTree[] | ReactElement[];
  node: ReactElement;
  props: ObjectMap;
  type: string;
}

export type TreeNode = SerializedTree;

export type ExtractedContexts = Map<React.ExoticComponent<React.ConsumerProps<any>>, any>; // tslint:disable-line no-any

export interface UnwrapElementResult {
  contexts: ExtractedContexts;
  element: ComponentTypeElement;
}

export interface SCSerializedTree {
  $$typeof: symbol;
  children: null;
  node: AnyStyledComponent;
  props: SCSerializedTreeProps;
  type: string;
}

export interface SCSerializedTreeProps {
  forwardedComponent: SCForwardRefComponent;
}

export type SCForwardRefComponent = ForwardRefComponent & { componentStyle: SCComponentStyle; displayName: string };

export interface SCComponentStyle {
  rules: Array<string | number | Func>;
}
