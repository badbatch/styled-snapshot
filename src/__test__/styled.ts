import styled from "styled-components";

export const StyledDiv = styled.div`
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  box-sizing: border-box;
  position: relative;
`;
