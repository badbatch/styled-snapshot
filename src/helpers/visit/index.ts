import { Func, ObjectMap } from "@repodog/types";
import { Json } from "enzyme-to-json";
import { castArray, isFunction, isPlainObject } from "lodash";
import { FunctionComponent, ReactElement, cloneElement } from "react";
import { ForwardRef, isElement, isPortal } from "react-is";
import { PORTAL, RENDER_PROP } from "../../constants";
import { ReactTreeVisitor, SerializedTree, TreeNode } from "../../types";
import createSnapshotElement from "../create-snapshot-element";
import getComponentName from "../get-component-name";
import isFunctionComponent from "../is-function-component";

export default function visit(serializedComponent: Json, visitor?: ReactTreeVisitor) {
  visitNode(serializedComponent as SerializedTree, visitor);
  return serializedComponent;
}

function visitChildren(children: ReactElement | ReactElement[], visitor?: ReactTreeVisitor) {
  castArray(children).forEach(child => {
    visitElement(child, visitor);
  });
}

function visitFunctionProp(val: Func | FunctionComponent, visitor?: ReactTreeVisitor) {
  if (isFunctionComponent(val)) {
    const component = val as FunctionComponent;
    return Symbol(getComponentName(component));
  } else {
    const func = val as Func;

    try {
      const output = func();
      return isElement(output) ? createSnapshotElement(RENDER_PROP, visitElement(output, visitor)) : val;
    } catch (error) {
      return val;
    }
  }
}

function visitNode(treeNode: TreeNode, visitor?: ReactTreeVisitor) {
  if (isFunction(visitor)) visitor(treeNode);

  if (treeNode.props) {
    visitProps(treeNode.props, visitor);
  }

  if (treeNode.children) {
    const children = treeNode.children as TreeNode[];

    children.forEach((child, index) => {
      treeNode.children[index] = visitElement(child.node, visitor);
    });
  }
}

function visitProps(props: ObjectMap, visitor?: ReactTreeVisitor) {
  Object.keys(props).forEach(key => {
    const val = props[key];

    switch (true) {
      case isFunction(val):
        props[key] = visitFunctionProp(val, visitor);
        break;
      case isElement(val):
        props[key] = visitElement(val, visitor);
        break;
      case isPortal(val):
        props[key] = createSnapshotElement(PORTAL, val.children);
        break;
      case isPlainObject(val) && val.$$typeof === ForwardRef:
        props[key] = Symbol(getComponentName(val));
      // no default
    }
  });

  if (props.children) {
    visitChildren(props.children, visitor);
  }
}

function visitElement(element: ReactElement, visitor?: ReactTreeVisitor) {
  const nodeClone = { props: { ...element.props } };
  if (isFunction(visitor)) visitor(nodeClone);
  visitProps(nodeClone.props, visitor);
  return cloneElement(element, nodeClone.props);
}
