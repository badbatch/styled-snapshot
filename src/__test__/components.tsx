import { ObjectMap } from "@repodog/types";
import PropTypes from "prop-types";
import React, { Component, ComponentType, forwardRef, memo } from "react";
import { createPortal } from "react-dom";
import styled, { ThemeContext, ThemeProvider, withTheme } from "styled-components";
import { StyledDiv, StyledItem, StyledList } from "./styled";
import { ComponentProps } from "./types";

export class ClassComponent extends Component<ComponentProps> {
  public static displayName = "ClassComponent";

  public static propTypes = {
    Component: PropTypes.elementType,
  };

  // tslint:disable-next-line prefer-function-over-method
  public render() {
    return this.props.children;
  }
}

export const ContextComponent = () => (
  <ThemeContext.Provider value="dark">
    <ContextConsumerComponent />
  </ThemeContext.Provider>
);

ContextComponent.displayName = "ContextComponent";

export const ContextConsumerComponent = () => (
  <ThemeContext.Consumer>
    {value => (
      <ClassComponent value={value}>
        <FunctionComponent />
      </ClassComponent>
    )}
  </ThemeContext.Consumer>
);

ContextConsumerComponent.displayName = "ContextConsumerComponent";

export const ContextProviderComponent = () => (
  <ThemeContext.Provider value="dark">
    <ClassComponent>
      <FunctionComponent />
    </ClassComponent>
  </ThemeContext.Provider>
);

ContextProviderComponent.displayName = "ContextProviderComponent";

export const ForwardRefComponent = forwardRef<HTMLDivElement, ComponentProps>((props, ref) => (
  <ClassComponent {...props} innerRef={ref}>
    <FunctionComponent />
  </ClassComponent>
));

ForwardRefComponent.displayName = "ForwardRefComponent";

export const FunctionComponent = (props: ComponentProps) => <StyledDiv {...props} />;
FunctionComponent.displayName = "FunctionComponent";

FunctionComponent.propTypes = {
  Component: PropTypes.elementType,
};

export const MemoComponent = memo(FunctionComponent);
MemoComponent.displayName = "MemoComponent";

export const portal = createPortal(
  <ClassComponent>
    <FunctionComponent />
  </ClassComponent>,
  document.createElement("div"),
);

export const PortalComponent = () => portal;
PortalComponent.displayName = "PortalComponent";

export const SCComponent = () => (
  <StyledDiv>
    <ClassComponent>
      <FunctionComponent />
    </ClassComponent>
    <FunctionComponent />
    <StyledList>
      <StyledItem>One</StyledItem>
      <StyledItem>Two</StyledItem>
      <StyledItem>Three</StyledItem>
      <StyledItem>Four</StyledItem>
    </StyledList>
  </StyledDiv>
);

SCComponent.displayName = "SCComponent";

export const SCContextComponent = () => (
  <ThemeProvider theme={{ type: "light" }}>
    <WithThemeClassComponent />
  </ThemeProvider>
);

SCContextComponent.displayName = "SCContextComponent";

export const WithThemeClassComponent = withTheme<ComponentType<any>>(ClassComponent); // tslint:disable-line no-any

export const SerializeComponent = (props: ObjectMap = {}) => {
  return (
    <ClassComponent
      Component={FunctionComponent}
      callback={() => null}
      consumer={<ThemeContext.Consumer>{() => <div />}</ThemeContext.Consumer>}
      decorator={<WithThemeClassComponent />}
      element={<FunctionComponent {...props} />}
      forwardRef={<ForwardRefComponent />}
      fragment={<></>}
      memo={<MemoComponent />}
      portal={portal}
      provider={<ThemeContext.Provider value="" />}
      renderProp={() => <FunctionComponent {...props} renderProp={subProps => <div {...subProps} />} />}
      styled={StyledDiv}
      styledFunctionComponent={<StyledFunctionComponent />}
      styledStyledComponent={<StyledStyledComponent />}
    >
      <ForwardRefComponent />
      <FunctionComponent {...props} />
      <MemoComponent />
      <StyledFunctionComponent />
      <StyledList>
        {[1, 2, 3, 4].map(id => (
          <StyledStyledComponent key={id} />
        ))}
      </StyledList>
    </ClassComponent>
  );
};

SerializeComponent.displayName = "SerializeComponent";

export const StyledFunctionComponent = styled(FunctionComponent)``;

export const StyledStyledComponent = styled(StyledDiv)``;
