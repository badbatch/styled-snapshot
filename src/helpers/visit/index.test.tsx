import { ObjectMap } from "@repodog/types";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import { has } from "lodash";
import React from "react";
import visit from ".";
import { SerializeComponent } from "../../__test__/components";
import { SerializedTree } from "../../types";

describe("visit", () => {
  describe("when run over a serialized component tree", () => {
    it("should return the correct structure", () => {
      const componentTree = shallow(<SerializeComponent />);
      const serializedTree = toJson(componentTree);
      expect(visit(serializedTree as SerializedTree, {})).toMatchSnapshot();
    });
  });

  describe("when a visitor is passed in to remove certain data", () => {
    it("should return the correct structure", () => {
      const componentTree = shallow(<SerializeComponent theme={{ type: "light" }} />);
      const serializedTree = toJson(componentTree);

      const reactTreeVisitor = (node: ObjectMap) => {
        if (has(node, ["props", "theme"])) {
          node.props.theme = null;
        }
      };

      expect(visit(serializedTree as SerializedTree, { reactTreeVisitor })).toMatchSnapshot();
    });
  });
});
