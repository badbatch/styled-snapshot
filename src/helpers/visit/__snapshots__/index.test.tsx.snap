// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`visit when a visitor is passed in to remove certain data should return the correct structure 1`] = `
<ClassComponent
  Component={Symbol(FunctionComponent)}
  callback={[Function]}
  consumer={
    <Context.Consumer>
      <RenderProp(div) />
    </Context.Consumer>
  }
  element={
    <FunctionComponent
      theme={null}
    />
  }
  forwardRef={<ForwardRef />}
  fragment={<React.Fragment />}
  memo={<Memo(MemoComponent) />}
  portal={<Portal(ClassComponent) />}
  provider={
    <Context.Provider
      value=""
    />
  }
  renderProp={
    <RenderProp(FunctionComponent)
      theme={null}
    />
  }
  styled={Symbol(styled.div)}
>
  <ForwardRef />
  <FunctionComponent
    theme={null}
  />
  <Memo(MemoComponent) />
</ClassComponent>
`;

exports[`visit when run over a serialized component tree should return the correct structure 1`] = `
<ClassComponent
  Component={Symbol(FunctionComponent)}
  callback={[Function]}
  consumer={
    <Context.Consumer>
      <RenderProp(div) />
    </Context.Consumer>
  }
  element={<FunctionComponent />}
  forwardRef={<ForwardRef />}
  fragment={<React.Fragment />}
  memo={<Memo(MemoComponent) />}
  portal={<Portal(ClassComponent) />}
  provider={
    <Context.Provider
      value=""
    />
  }
  renderProp={<RenderProp(FunctionComponent) />}
  styled={Symbol(styled.div)}
>
  <ForwardRef />
  <FunctionComponent />
  <Memo(MemoComponent) />
</ClassComponent>
`;
